#1589 - Rejection Sampling

ueliwechsler opened the pull request:
Closes #1589.

Not anything is documented yet because I am not sure about the general structure. 
I created a new type `Sampler`, in order to have more flexibility and also use the Sampler outside of LazySets without recomputing the `Sampler.sampler` variable again ... maybe you have an idea for a better name there ;)
I am looking forward to your feedback.

schillic commented on src/sample_from_set.jl:
```diff
@@ -0,0 +1,100 @@
+
+A vector of `n_samples` vectors.
+"""
+function Base.rand(P::LazySet{N}, n_samples::Int;
```
@mforets does not like `rand`.

ueliwechsler replied:
I like the use of `rand` in julia in general because then you don't have to think about what function to use to get an instance or a sample of a specific type. But I am open to suggestions. Would `sample` better?

mforets replied:
Indeed, we had discussion about this and the functions are so different that keeping the names separate is better IMO, provided that:

- `rand` receives a set *type* and returns a LazySet instance -- (i.e. an infinite number of points, and you draw from the distribution on the type given some constraints...)
- `sample` receives a LazySet *instance* and returns a sample of points in the set -- (a finite number of points and you draw from a distribution which is eg. uniform over the set)

ueliwechsler replied:
👍
Should I import `sample` from `Distributions`? Because I got the warning message
```WARNING: both Distributions and LazySets export "sample"; uses of it in module Main must be qualified```

schillic replied:
Yes, you should.

ueliwechsler replied:
I did this. But there is still the warning.
`WARNING: import of Distributions.sample into LazySets conflicts with an existing identifier; ignored.`
I think this is due to the require in https://github.com/JuliaReach/LazySets.jl/blob/f3e5fbadee35a13c26199d72f289055ad89bf15c/src/Ball2.jl#L266

Should we move the sample function outside of the `samples.jl` file?

schillic replied:
I found that just doing `import .Distributions` instead of `import .Distributions: sample` works.

schillic commented on src/sample_from_set.jl:
```diff
@@ -0,0 +1,100 @@
+function load_distributions_sample_from_set()
```
Maybe the file name `sample_from_set.jl` could be shortened to `sample.jl`. @mforets?

ueliwechsler replied:
There is also a file `samples.jl` should I merge this to files?

schillic replied:
Yes!

mforets commented on src/sample_from_set.jl:
```diff
@@ -0,0 +1,100 @@
+        x[j,:] = [-ρ(-ej, P), ρ(ej, P)]
+    end
+    return x
+end
```
Well it depends, box approximation can (and should) dispatch on the type of X hence using it is preferable. For instance, if X is an hyperrectangle the operation should be a no-op.

schillic replied:
True, I was only considering the creation of the matrix via `hcat` and the fact that our `Hyperrectangle` has a different representation, so it has to compute center/radius and convert back. But yes, this is in the order of milliseconds, so efficiency is really not an argument here.

ueliwechsler replied:
Okay. Where should this function be located?

schillic replied:
It could just replace your function (remove the `2` in the name of course).

ueliwechsler replied:
Sorry, I was unclear. I was just wondering if this new function should be in the `samples.jl`  file or somewhere else.

schillic replied:
Ah. It is a general function, but we never needed it so far. The name is not clear for a general purpose (but I cannot think of a good name). So I would
1. either put it in `samples.jl` (and maybe add a leading `_` which we use to mark internal functions)
2. or put it in `AbstractHyperrectangle.jl` if you find a name that clearly indicates what this function does.

ueliwechsler replied:
Okay. I will go with the first version.

Just as a little caveat. With the new function the test for an unbounded `Polyhedra` breaks. 

```julia
P3 = HPolyhedron([0.0 0.1], [3.0])
@show box_approximation(P3)
@test LazySets.canonical_length(P3) == [-Inf Inf; -Inf 30.0]
```
```julia
a = box_approximation(P3) = Hyperrectangle{Float64}([NaN, -Inf], [Inf, Inf])
Evaluated: [NaN NaN; -Inf NaN] == [-Inf Inf; -Inf 30.0]
```

schillic replied:
Oh, true! Then you should of course keep your version :+1:

schillic commented on src/sample_from_set.jl:
```diff
@@ -0,0 +1,9 @@
+function load_distributions_sample_from_set()
```
This file can now be removed.

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+
+function Sampler(X, distribution=Uniform)
+    box_lengths = _canonical_length(X)
+    sampler = [distribution(length...) for length in eachrow(box_lengths)]
```
This throws `LoadError: UndefVarError: eachrow not defined` in Julia v1.0.

ueliwechsler replied:
It should work now.

schillic commented on test/runtests.jl:
```diff
@@ -143,6 +143,11 @@ if test_suite_basic
     include("check_method_implementation.jl")
     @time @testset "LazySets.interfaces" begin include("unit_interfaces.jl") end
+
+    # Also loads Distributions
```
```suggestion
```

schillic commented on test/unit_samples.jl:
```diff
@@ -0,0 +1,31 @@
+
+using LazySets
+using Distributions
+using Polyhedra
```
Eventually remove these lines above (they are not needed).

schillic replied:
There are some commented lines on top that can also be removed.

ueliwechsler replied:
Will do that. I left them there because I am still not able to run the `runtest.jl` file due to problem with `CDDLib` on Windows.

schillic commented on src/samples.jl:
```diff
@@ -1,7 +1,7 @@
 return quote
 
-using .Distributions: Uniform, Normal
+using .Distributions: Distribution, Uniform, Normal
```
You never used `Distribution`... :confused:

ueliwechsler replied:
Right, in the beginning I had
```julia
mutable struct Sampler 
    X::LazySet
    sampler::Vector{<:Distribution}
end
```
Should I add this? What is best practice?

schillic replied:
I think the best is to add type parameters.
```julia
struct Sampler{S<:LazySet, D<:Distribution}
    X::S
    sampler::Vector{D}
end
```
Note that the struct needs not be `mutable` if you only ever modify the `Vector`'s contents.

mforets replied:
> What is best practice?

Indeed, for library code it is preferable to consider using type parameters; see the discussion in https://docs.julialang.org/en/v1/manual/performance-tips/#Type-declarations-1

ueliwechsler replied:
Thanks for the explanation. I will add your version.

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+
+### Inputs
+
+- `X`   - lazyset
```
```suggestion
- `X` - convex set
```

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+
+### Outputs
+
+Matrix with two columns and `n=dims(X)` rows. Each column stands for
```
```suggestion
A matrix with `n = dims(X)` rows and two columns. Each row stands for
```

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+
+Matrix with two columns and `n=dims(X)` rows. Each column stands for
+one dimension of `X` whereas the first column is the minimum and the second
+column the maximum value of the corresponding dimension.
```
```suggestion
column is the maximum value of the corresponding dimension.
```

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+function _canonical_length(X::LazySet{N}) where {N<:Real}
+    dims = dim(X)
+    x = Matrix{N}(undef, dims, 2)
+    for j=1:dims
```
```suggestion
    for j = 1:dims
```

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+           rng::AbstractRNG=GLOBAL_RNG,
+           seed::Union{Int, Nothing}=nothing) where {N}
+
+Rejection sampling of an arbitrary LazySet `X` for which the support value
```
```suggestion
Rejection sampling of an arbitrary `LazySet` `X`.
```
(The support function is a requirement of this library.)

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+           seed::Union{Int, Nothing}=nothing) where {N}
+
+Rejection sampling of an arbitrary LazySet `X` for which the support value
+function is defined.
```
```suggestion
```

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+
+Draw a sample `x` from a uniform distribution of a box-overapproximation of the
+original set `X` in all `n` dimensions. The function rejects a drawn sample `x`
+and redraws as long as the sample is not contained in the original set `X`,
```
Move this paragraph to a new section `### Algorithm` at the end of this docstring.

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+### Input
+
+- `X`           -- lazyset
+- `n_samples`    -- number of random samples
```
```suggestion
- `n_samples`   -- number of random samples
```

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+
+### Input
+
+- `X`           -- lazyset
```
```suggestion
- `X`           -- convex set
```

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+                rng::AbstractRNG=GLOBAL_RNG,
+                seed::Union{Int, Nothing}=nothing) where {N<:Real}
+    rng = reseed(rng, seed)
+    @assert isbounded(X) "this function requires that the set `X` is bounded, but it is not"
```
Huh, why should the set be bounded?

mforets replied:
Rejection sampling proceeds by taking a box overapproximation of the set, and sample inside that box. If the set is unbounded, the box will have infinities. So in that case we  should sample from a distribution with unbounded support. It's ok to restrict to bounded sets in this issue (and open a new issue for unbounded sets so we don't forget).

schillic replied:
I was referring to [this comment](https://github.com/JuliaReach/LazySets.jl/pull/1612#discussion_r314941390). The only reason why we kept the function `_canonical_length` was because it additionally supports unbounded sets. And this assertion makes that case break.

Note that `rand` is never tested for unbounded sets here (and I bet it would fail).

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+    return rand(X,1)[1]
+end
+
+mutable struct Sampler
```
Maybe call it `BoxSampler`?

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+mutable struct Sampler
+    X
+    sampler
+    distribution
```
`distribution` is actually never used. This does not necessarily mean that it is not good to store this information, though.

ueliwechsler replied:
Right. But I think, it is not necessary. The distribution information is already stored in `sampler`.

schillic commented on src/samples.jl:
```diff
@@ -117,5 +117,100 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+               sampler::Sampler;
+               rng::AbstractRNG=GLOBAL_RNG) where {N}
+    @inbounds for i in 1:length(D)
+        while true
```
Just a comment: I find the following loop easier to read:
```julia
i = 0
n = @inbounds while i < length(D)
    w = rand.(Ref(rng), sampler.sampler)
    if w ∈ sampler.X
        i += 1
        D[i] = w
    end
end
```

ueliwechsler replied:
What is the reason for using `n = `?

schillic replied:
The is a relict of a previous version where I precomputed `length(D)`, but I decided that this is not worth adding a line. Just remove it :wink:

ueliwechsler replied:
For me both versions are not that readable, to be honest ;)
I will add a clearer but slightly more verbose version and like to hear your opinion.

schillic replied:
Looks good!

schillic commented on test/unit_samples.jl:
```diff
@@ -0,0 +1,31 @@
+using Distributions
+using Polyhedra
+
+n = 5
```
```suggestion
```

schillic commented on test/unit_samples.jl:
```diff
@@ -0,0 +1,31 @@
+b = [1.0, 2.0, 3.0, 4.0]
+P2 = HPolyhedron(A, b)
+P3 = HPolyhedron([0.0 0.1], [3.0])
+P4 = rand(Ball2, dim=n)
```
```suggestion
P4 = Ball2{Float64}([0.2, -0.3, -1.1, 0.6, -0.7], 0.4)
```
I prefer to have deterministic tests where possible for reproducibility.
(Of course the sampling itself can be random.)

schillic commented on test/unit_samples.jl:
```diff
@@ -0,0 +1,31 @@
+using Polyhedra
+
+n = 5
+P1 = rand(BallInf, dim=n)
```
```suggestion
P1 = BallInf{Float64}([0.6, 0.1, -1.3, -0.4, 0.2], 0.6)
```

schillic commented on test/unit_samples.jl:
```diff
@@ -0,0 +1,28 @@
+@test LazySets._canonical_length(P3) == [-Inf Inf; -Inf 30.0]
+@test LazySets._canonical_length(P4+-P4.center) ≈ radius(P4)*[-ones(5) ones(5)]
+
+@test LazySets.SetSampler(P2).sampler == [Uniform(-3.0,1.0), Uniform(-4.0,2.0)]
```
You need to add `using Distributions: Uniform` to the top of this file.

mforets commented on src/samples.jl:
```diff
@@ -117,5 +117,98 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+"""
+    _canonical_length(X::LazySet)
+
+Get the canonical length of the box-approximation of the convex set `X`.
```
I had to look at the code to understand the meaning of "canonical length"; maybe say that the function `Returns the support function of the given set along the positive and negative canonical directions.`

mforets commented on src/samples.jl:
```diff
@@ -117,5 +117,98 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+
+A vector of `n_samples` vectors.
+"""
+function sample(X::LazySet{N}, n_samples::Int;
```
Let's call the second argument `num_samples` (compare to `num_vertices`, or `num_constraints`).

schillic replied:
@mforets: Just a comment: in https://github.com/JuliaReach/LazySets.jl/blob/c4e52db11b6c90543026a34d08571e7ae537bdf5/src/samples.jl#L98 you called the variable `p` (which I find arbitrary).

mforets commented on src/samples.jl:
```diff
@@ -117,5 +117,98 @@ function _sample_unit_nball_muller!(D::Vector{Vector{N}}, n::Int, p::Int;
+    return D
+end
+
+function sample(X::LazySet{N}; kwargs...) where {N<:Real}
```
This is ok but i don't see a compelling reason to not give the default value, 1, to the function defined in line 171. The second argument can still have a default value without being a keyword argument. What do you think?

ueliwechsler replied:
I agree with you in general. But I did this to avoid having a `Vector{Vector}` as a result when I only want one sample from a set. Therefore the `[1]` in `sample(X, 1; kwargs...)[1]`. Is there a better way to handle this?

schillic replied:
I disagree. Type instabilities are not good.

schillic replied:
You can add the default value `1` to the other method, but the return type should *always* be a vector of vectors. For convenience it is useful to have this other method that only returns a single vector instead.

schillic replied:
True, then forget about the default value `1`.

schillic replied:
I misunderstood your comment completely, sorry. I agree with what you said :+1:

ueliwechsler commented:
I added some more documentation and changed some function names. I hope this does not make it worse. Is there anything to do? @schillic I think you were also saying, that it does not make to use `_canonical_length` when the only case it would be "better" than the `box_approximation` version is momentary not support (because `rand(Uniform(Inf,3))` does not work).

schillic commented:
> Is there anything to do?

* There are still some comments to resolve (they are hidden because there are so many; you need to click "Load more") [here](https://github.com/JuliaReach/LazySets.jl/pull/1612#discussion_r314948323), [here](https://github.com/JuliaReach/LazySets.jl/pull/1612#discussion_r314948329), [here](https://github.com/JuliaReach/LazySets.jl/pull/1612#discussion_r314948368), and [here](https://github.com/JuliaReach/LazySets.jl/pull/1612#discussion_r314948387).
* I reopened a discussion [here](https://github.com/JuliaReach/LazySets.jl/pull/1612#discussion_r314948444) (the name `Sampler` sounds general, but the implementation only works for boxes; if we ever plan to add other samplers, we would have to rename this type).
* You need to fix the merge conflicts (for instance by rebasing with master: `git rebase master`).
* After that, we do the next iteration :repeat: :smiling_imp:

schillic commented:
> I think you were also saying, that it does not make to use _canonical_length when the only case it would be "better" than the box_approximation version is momentary not support (because rand(Uniform(Inf,3)) does not work).

True, I would prefer to remove this function (mainly because it is unclear from the name what it does and we might get confused about it).

mforets commented:
> You need to fix the merge conflicts (for instance by rebasing with master: git rebase master).

Conflicts resolved.

mforets commented:
@ueliwechsler  Thanks again for contributing with this PR that adds rejection sampling for an arbitrary lazyset 🎉 👍  The PR looks good and is functional, and is covered with unit tests. Let me address the remaining review comments in a follow-up PR.
